include/dlaf/communication/executor.h

```diff
@@ -0,0 +1,162 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2021, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+/// @file
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+#include <utility>
+
+#include <hpx/async_mpi/mpi_future.hpp>
+#include <hpx/datastructures/tuple.hpp>
+#include <hpx/execution.hpp>
+#include <hpx/functional.hpp>
+#include <hpx/future.hpp>
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/synchronization/mutex.hpp>
+#include <hpx/type_support/unused.hpp>
+
+#include <mpi.h>
+
+#include "dlaf/common/assert.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/init.h"
+#include "dlaf/communication/mech.h"
+
+namespace dlaf {
+namespace comm {
+namespace detail {
+
+// Requests are only handled for the non-blocking version of comm::Executor
+template <MPIMech mech>
+struct request_handler {};
+
+inline void handle_request(MPIMech mech, MPI_Request req) {
+  if (mech == MPIMech::Yielding) {
+    hpx::util::yield_while([&req] {
+      int flag;
+      mpi_invoke(MPI_Test, &req, &flag, MPI_STATUS_IGNORE);
+      return flag == 0;
+    });
+  }
+  else if (mech == MPIMech::Polling) {
+    hpx::mpi::experimental::get_future(req).get();
+  }
+  else {
+    std::cout << "UNIMPLEMENTED!" << std::endl;
```

Comment:
very minor: you can use `DLAF_UNIMPLEMENTED`

Reply:
Very minor tangent: would it be useful to have a `DLAF_UNREACHABLE` as well?